File: src/logger.ts
/**
 * Internal Logger for TraceFlow SDK
 * Configurable logging system with level-based filtering
 */

type LogLevel = 'debug' | 'info' | 'warn' | 'error';

const LOG_LEVELS: Record<LogLevel, number> = {
  debug: 0,
  info: 1,
  warn: 2,
  error: 3,
};

export interface LoggerConfig {
  enabled?: boolean;
  minLevel?: LogLevel;
  customLogger?: {
    debug: (message: string, ...args: any[]) => void;
    info: (message: string, ...args: any[]) => void;
    warn: (message: string, ...args: any[]) => void;
    error: (message: string, ...args: any[]) => void;
  };
}

/**
 * Logger class for SDK internal logging
 */
export class Logger {
  private enabled: boolean;
  private minLevel: number;
  private customLogger?: LoggerConfig['customLogger'];

  constructor(config: LoggerConfig = {}) {
    this.enabled = config.enabled ?? true;
    this.minLevel = LOG_LEVELS[config.minLevel || 'info'];
    this.customLogger = config.customLogger;
  }

  /**
   * Log debug message
   */
  debug(message: string, ...args: any[]): void {
    if (!this.enabled || this.minLevel > LOG_LEVELS.debug) return;
    
    if (this.customLogger?.debug) {
      this.customLogger.debug(message, ...args);
    } else {
      console.debug(`[TraceFlow:DEBUG] ${message}`, ...args);
    }
  }

  /**
   * Log info message
   */
  info(message: string, ...args: any[]): void {
    if (!this.enabled || this.minLevel > LOG_LEVELS.info) return;
    
    if (this.customLogger?.info) {
      this.customLogger.info(message, ...args);
    } else {
      console.log(`[TraceFlow:INFO] ${message}`, ...args);
    }
  }

  /**
   * Log warning message
   */
  warn(message: string, ...args: any[]): void {
    if (!this.enabled || this.minLevel > LOG_LEVELS.warn) return;
    
    if (this.customLogger?.warn) {
      this.customLogger.warn(message, ...args);
    } else {
      console.warn(`[TraceFlow:WARN] ${message}`, ...args);
    }
  }

  /**
   * Log error message
   */
  error(message: string, ...args: any[]): void {
    if (!this.enabled || this.minLevel > LOG_LEVELS.error) return;
    
    if (this.customLogger?.error) {
      this.customLogger.error(message, ...args);
    } else {
      console.error(`[TraceFlow:ERROR] ${message}`, ...args);
    }
  }

  /**
   * Create a scoped logger with a prefix
   */
  scope(prefix: string): ScopedLogger {
    return new ScopedLogger(this, prefix);
  }
}

/**
 * Scoped logger with automatic prefix
 */
class ScopedLogger {
  constructor(
    private logger: Logger,
    private prefix: string
  ) {}

  debug(message: string, ...args: any[]): void {
    this.logger.debug(`${this.prefix} ${message}`, ...args);
  }

  info(message: string, ...args: any[]): void {
    this.logger.info(`${this.prefix} ${message}`, ...args);
  }

  warn(message: string, ...args: any[]): void {
    this.logger.warn(`${this.prefix} ${message}`, ...args);
  }

  error(message: string, ...args: any[]): void {
    this.logger.error(`${this.prefix} ${message}`, ...args);
  }
}

File: go/logger.go
package traceflow

import (
	"fmt"
	"log"
	"strings"
)

// Logger is the minimal logging surface used internally by the SDK. Provide a
// custom implementation via Config.Logger to integrate with your logging stack.
type Logger interface {
	Debug(format string, args ...any)
	Info(format string, args ...any)
	Warn(format string, args ...any)
	Error(format string, args ...any)
}

type logLevel int

const (
	levelDebug logLevel = iota
	levelInfo
	levelWarn
	levelError
)

func parseLevel(s string) logLevel {
	switch strings.ToLower(s) {
	case "debug":
		return levelDebug
	case "warn":
		return levelWarn
	case "error":
		return levelError
	default:
		return levelInfo
	}
}

// stdLogger is the default Logger; it writes level-prefixed lines to the
// standard library logger and filters by minimum level.
type stdLogger struct {
	enabled  bool
	minLevel logLevel
	prefix   string
}

func newStdLogger(enabled bool, minLevel string) *stdLogger {
	return &stdLogger{enabled: enabled, minLevel: parseLevel(minLevel)}
}

func (l *stdLogger) emit(level logLevel, tag, format string, args ...any) {
	if !l.enabled || level < l.minLevel {
		return
	}
	msg := fmt.Sprintf(format, args...)
	if l.prefix != "" {
		log.Printf("[TraceFlow:%s] %s %s", tag, l.prefix, msg)
		return
	}
	log.Printf("[TraceFlow:%s] %s", tag, msg)
}

func (l *stdLogger) Debug(format string, args ...any) { l.emit(levelDebug, "DEBUG", format, args...) }
func (l *stdLogger) Info(format string, args ...any)  { l.emit(levelInfo, "INFO", format, args...) }
func (l *stdLogger) Warn(format string, args ...any)  { l.emit(levelWarn, "WARN", format, args...) }
func (l *stdLogger) Error(format string, args ...any) { l.emit(levelError, "ERROR", format, args...) }

// scope returns a logger that prefixes every message, mirroring the JS SDK's
// scoped loggers (e.g. "HTTP").
func (l *stdLogger) scope(prefix string) *stdLogger {
	clone := *l
	clone.prefix = strings.TrimSpace(l.prefix + " " + prefix)
	return &clone
}

// noopLogger discards everything; used as a safe fallback.
type noopLogger struct{}

func (noopLogger) Debug(string, ...any) {}
func (noopLogger) Info(string, ...any)  {}
func (noopLogger) Warn(string, ...any)  {}
func (noopLogger) Error(string, ...any) {}
